src/jit/lower.cpp

```diff
@@ -1922,251 +1922,438 @@ GenTree* Lowering::LowerTailCallViaHelper(GenTreeCall* call, GenTree* callTarget
+                    break;
+                case GT_GE:
+                    cmp->gtOper = GT_LT;
+                    hiCmpOper   = GT_GT;
+                    loCmpOper   = GT_GE;
+                    break;
+                default:
+                    unreached();
+            }
+
+            BasicBlock* newBlock2 = comp->fgSplitBlockAtEnd(newBlock);
+
+            GenTree* hiJcc = new (comp, GT_JCC) GenTreeJumpCC(hiCmpOper);
+            hiJcc->gtFlags = cmp->gtFlags;
+            LIR::AsRange(newBlock).InsertAfter(nullptr, hiJcc);
+
+            BlockRange().Remove(loSrc1.Def());
+            BlockRange().Remove(loSrc2.Def());
+            GenTree* loCmp   = comp->gtNewOperNode(loCmpOper, TYP_INT, loSrc1.Def(), loSrc2.Def());
+            loCmp->gtFlags   = cmp->gtFlags | GTF_UNSIGNED;
+            GenTree* loJtrue = comp->gtNewOperNode(GT_JTRUE, TYP_VOID, loCmp);
+            LIR::AsRange(newBlock2).InsertAfter(nullptr, loSrc1.Def(), loSrc2.Def(), loCmp, loJtrue);
+
+            m_block->bbJumpKind = BBJ_COND;
             m_block->bbJumpDest = nextDest;
             nextDest->bbFlags |= BBF_JMP_TARGET;
             comp->fgAddRefPred(nextDest, m_block);
+
+            newBlock->bbJumpKind = BBJ_COND;
+            newBlock->bbJumpDest = jumpDest;
+            comp->fgAddRefPred(jumpDest, newBlock);
+
+            assert(newBlock2->bbJumpKind == BBJ_COND);
+            assert(newBlock2->bbJumpDest == jumpDest);
         }
-        else
+
+        BlockRange().Remove(src1);
+        BlockRange().Remove(src2);
+    }
+#endif
+
+#ifdef _TARGET_XARCH_
+#ifdef _TARGET_AMD64_
+    if (cmp->gtGetOp1()->TypeGet() != cmp->gtGetOp2()->TypeGet())
+    {
+        bool op1Is64Bit = (genTypeSize(cmp->gtGetOp1()->TypeGet()) == 8);
+        bool op2Is64Bit = (genTypeSize(cmp->gtGetOp2()->TypeGet()) == 8);
+
+        if (op1Is64Bit != op2Is64Bit)
         {
-            m_block->bbJumpDest = jumpDest;
-            comp->fgAddRefPred(jumpDest, m_block);
-        }
+            //
+            // Normally this should not happen. IL allows comparing int32 to native int but the importer
+            // automatically inserts a cast from int32 to long on 64 bit architectures. However, the JIT
+            // accidentally generates int/long comparisons internally:
+            //   - loop cloning compares int (and even small int) index limits against long constants
+            //   - switch lowering compares a 64 bit switch value against a int32 constant
+            //
+            // TODO-Cleanup: The above mentioned issues should be fixed and then the code below may be
+            // replaced with an assert or at least simplified. The special casing of constants in code
+            // below is only necessary to prevent worse code generation for switches and loop cloning.
+            //
+
+            GenTree*& longOp    = op1Is64Bit ? cmp->gtOp.gtOp1 : cmp->gtOp.gtOp2;
+            GenTree*& smallerOp = op2Is64Bit ? cmp->gtOp.gtOp1 : cmp->gtOp.gtOp2;
```

Comment:
Nit: we generally prefer to use pointers instead of references when the backing storage is going to be modified, as it makes the modification clearer at the point at which it occurs (e.g. line 2217 below).